proxy/src/cache.rs

```diff
@@ -0,0 +1,221 @@
+use std::{
+    borrow::Borrow,
+    hash::Hash,
+    ops::{Deref, DerefMut},
+    time::{Duration, Instant},
+};
+
+// This seems to make more sense than `lru` or `cached`:
+//
+// * `near/nearcore` ditched `cached` in favor of `lru`
+//   (https://github.com/near/nearcore/issues?q=is%3Aissue+lru+is%3Aclosed).
+//
+// * `lru` methods use an obscure `KeyRef` type in their contraints (which is deliberately excluded from docs).
+//   This severely hinders its usage both in terms of creating wrappers and supported key types.
+//
+// On the other hand, `hashlink` has good download stats and appears to be maintained.
+use hashlink::{linked_hash_map::RawEntryMut, LruCache};
+
+pub trait Cache {
+    /// Entry's key.
+    type Key;
+    /// Entry's value.
+    type Value;
+    /// Used for entry invalidation.
+    type LookupInfo<Key>;
+
+    /// Invalidate an entry using a lookup info.
+    fn invalidate(&self, _: &Self::LookupInfo<Self::Key>) {}
+}
```

Comment:
I think this trait is an unnecessary abstraction. There is only one implementation, TimedLru, and it makes it harder to understand.